OrbitVulkanLayer/SubmissionTracker.h

```diff
@@ -0,0 +1,757 @@
+// `TimerQueryPoool`. Note that the begin is optional, as it might not be part of the capture.
+// This struct is created if we capture the submission, so if we would not have captured the end
+// we also not have captured the begin, and don't need to store info about that command buffer at
+// all.
+struct SubmittedCommandBuffer {
+  std::optional<uint32_t> command_buffer_begin_slot_index;
+  uint32_t command_buffer_end_slot_index;
+};
+
+// A persistent version of a debug marker (either begin or end) that was submitted and its slot
+// in the `TimerQueryPool`. Note that we also store the `SubmissionMetaInformation` as begin markers
+// can originate in different submissions then the matching end markers.
+struct SubmittedMarker {
+  SubmissionMetaInformation meta_information;
+  uint32_t slot_index;
+};
+
+// Represents a color to be used to in debug markers. The values are all in range [0.f, 1.f.].
+struct Color {
+  float red;
+  float green;
+  float blue;
+  float alpha;
+};
+
+// Identifies a particular debug marker region. We have a stack of all markers of a queue, that gets
+// updated upon a submission (VkQueueSubmit). If at that time we have a value for the end_info, we
+// use that structure to persist the marker description. So once it is submitted, the
+// end_info will always set.
+// Beside the information about the begin/end, it also stores the text, color and the depth of the
+// marker. If a debug marker begin was limited because of its depth, cut_off is set to true.
+// This allows end markers on a different submission to also through the end marker away.
+// Example: Max Depth = 1
+// Submission 1: Begin("Foo"), Begin("Bar) -- "Bar" will be cut-off
+// Submission 2: End("Bar"), End("Foo") -- We now know, that the first end needs to be cut-off.
+struct MarkerState {
+  std::optional<SubmittedMarker> begin_info;
+  std::optional<SubmittedMarker> end_info;
+  std::string text;
+  Color color;
+  size_t depth;
+  bool cut_off;
+};
+
+// A single submission (VkQueueSubmit) can contain multiple `SubmitInfo
+struct SubmitInfo {
+  std::vector<SubmittedCommandBuffer> command_buffers;
+};
+
+// Wraps up all the data that needs to be persistent upon a submission (VkQueueSubmit).
+// `completed_markers` are all the debug markers, that got completed (via "End") within this
+// submission. Their "Begin" might still be in a different submission.
+struct QueueSubmission {
+  SubmissionMetaInformation meta_information;
+  std::vector<SubmitInfo> submit_infos;
+  std::vector<MarkerState> completed_markers;
+  uint32_t num_begin_markers = 0;
+};
+
+}  // namespace internal
+
+/*
+ * This class ultimately is responsible to track command buffer and debug marker timings.
+ * To do so, it keeps tracks of command-buffer allocations, destruction, begins, ends as well as
+ * submissions.
+ * On `VkBeginCommandBuffer` and `VkEndCommandBuffer` it can (if capturing) insert write timestamp
```

Comment:
"it can" or "it will"? If "if can" what are the cases when it will not insert write timestamps? I thought only for debug labels if the depth exceeds the desired depth, we would not insert write timestamps.

Reply:
The condition is "if we are capturing". I rephrased it slightly.